packages/toolpad-app/src/toolpad/propertyControls/DataProviderSelector.tsx

```diff
@@ -0,0 +1,331 @@
+              row
+              aria-labelledby={paginationModeSelectId}
+              value={options.paginationMode}
+              onChange={(event) =>
+                setOptions((existing) => ({
+                  ...existing,
+                  paginationMode: event.target.value as PaginationMode,
+                }))
+              }
+            >
+              <FormControlLabel value="index" control={<Radio />} label="Index based" />
+              <FormControlLabel value="cursor" control={<Radio />} label="Cursor based" />
+            </RadioGroup>
+          </FormControl>
+        </DialogContent>
+        <DialogActions>
+          <Button onClick={onClose}>Cancel</Button>
+          <LoadingButton
+            type="submit"
+            disabled={!newName || !!errorMessage}
+            loading={createProviderMutation.isLoading}
+          >
+            Create
+          </LoadingButton>
+        </DialogActions>
+      </form>
+    </Dialog>
+  );
+}
+
+function getProviderNameFromFile(file: FileIntrospectionResult): string {
+  return file.name.replace(/\.[^.]+$/, '');
+}
+
+function DataProviderSelector({ value, onChange }: EditorProps<string>) {
+  const { data: introspection, isLoading, error } = useFunctionsIntrospectQuery();
+
+  const options = React.useMemo<DataProviderSelectorOption[]>(() => {
+    return (
+      introspection?.files.flatMap((file) =>
+        file.dataProviders
+          ?.filter((dataProvider) => dataProvider.name === 'default')
+          .map((dataProvider) => ({
+            kind: 'option',
+            file,
+            dataProvider,
+            displayName: getProviderNameFromFile(file),
+          })),
+      ) ?? []
+    );
+  }, [introspection]);
+
+  const [fileName = null, providerName = null] = value ? value.split(':') : [];
+
+  const autocompleteValue = React.useMemo(() => {
+    return (
+      options.find(
+        (option) =>
+          option.kind === 'option' &&
+          option.file.name === fileName &&
+          option.dataProvider.name === providerName,
+      ) ?? null
+    );
+  }, [fileName, providerName, options]);
+
+  const errorMessage = error ? errorFrom(error).message : undefined;
+
+  const [dialogOpen, setDialogOpen] = React.useState(false);
+  const handleClose = () => {
+    setDialogOpen(false);
+  };
+  const [dialogValue, setDialogValue] = React.useState('');
+
+  const existingNames = React.useMemo(
+    () => new Set(introspection?.files.map((file) => getProviderNameFromFile(file))),
+    [introspection],
+  );
+
+  const handleCreateNewDataProvider = React.useCallback((suggestion: string) => {
+    setDialogValue(suggestion);
+    setDialogOpen(true);
+  }, []);
+
+  return (
+    <DataProviderSelectorRoot>
+      <CreateNewDataProviderDialog
+        open={dialogOpen}
+        onClose={handleClose}
+        onCommit={(newName) => onChange(`${newName}.ts:default`)}
+        initialName={dialogValue}
+        existingNames={existingNames}
```

Comment:
Not sure if it would be better to pass a validation function instead, anyway not a big deal.